totalRP3_Extended/aura/aura.lua

```diff
@@ -0,0 +1,1009 @@
+			end
+		end
+		return count;
+	end,
+	
+	IsAuraActive = function(self, auraId)
+		return self:FindAura(auraId) ~= nil;
+	end,
+	
+	isUpdating = false,
+	
+	Update = function(self, forceHardRefresh)
+		if self.isUpdating then return end
+		
+		self.isUpdating = true
+		if self.timer then
+			self.timer:Cancel();
+		end
+		
+		local now = self:Now();
+		local needsHardRefresh = forceHardRefresh;
+		local aurasRemoved = false;
+		
+		local updateCycle = 1;
+		
+		while updateCycle <= MAX_AURA_UPDATE_CYCLES do
+			
+			local scriptList = {};
+			
+			-- part 1: determine workflows that need to run
+			for i = 1,#self.activeAuras do
+				local aura = self.activeAuras[i];
+				if not aura.persistent.invalid then
+					
+					if aura.initialize then
+						if aura.class.LI and aura.class.LI.OA then
+							tinsert(scriptList, {
+								aura.class.LI.OA,
+								aura.class.SC or {},
+								{ object = aura.persistent },
+								aura.persistent.id
+							});
+						end
+						needsHardRefresh = true;
+						aura.initialize = nil;
+					end
+
+					if aura.class.BA.IV and aura.class.BA.IV < math.huge then
+						local nextTick = aura.persistent.lastTick + math.abs(aura.class.BA.IV);
+						if nextTick <= now then
+							aura.persistent.lastTick = nextTick;
+							if aura.class.LI and aura.class.LI.OT then
+								tinsert(scriptList, {
+									aura.class.LI.OT,
+									aura.class.SC or {},
+									{ object = aura.persistent },
+									aura.persistent.id
+								});
+							end
+						end
+					end
+					
+					if aura.cancel then
+						aura.cancel = nil;
+						aura.cancelled = true;
+						if aura.class.LI and aura.class.LI.OC then
+							tinsert(scriptList, {
+								aura.class.LI.OC,
+								aura.class.SC or {},
+								{ object = aura.persistent },
+								aura.persistent.id
+							});
+						end
+					end
+					
+					if aura.persistent.expiry <= now then
+						aura.expired = true;
+						if aura.class.LI and aura.class.LI.OE then
+							tinsert(scriptList, {
+								aura.class.LI.OE,
+								aura.class.SC or {},
+								{ object = aura.persistent },
+								aura.persistent.id
+							});
+						end
+					end
+				end
+			end
+			
+			-- part 2: run required workflows
+			local numScripts = #scriptList
+			for i = 1,numScripts do
+				TRP3_API.script.executeClassScript(scriptList[i][1], scriptList[i][2], scriptList[i][3], scriptList[i][4]);
+			end
+			
+			-- part 3: invalidate auras
+			for i = 1,#self.activeAuras do
+				if not self.activeAuras[i].persistent.invalid then
+					if self.activeAuras[i].cancelled or self.activeAuras[i].expired then
+						self.activeAuras[i].persistent.invalid = true;
+						self:UnregisterAuraEvents(self.activeAuras[i]);
+						aurasRemoved = true;
+					end
+				else
+					aurasRemoved = true;
+				end
+			end
+			
+			-- part 4: do it again in case there were workflows running
+			if numScripts <= 0 then
+				updateCycle = MAX_AURA_UPDATE_CYCLES + 1; -- get out of the loop
+			end
```

Comment:
I still don't really like this "repeat up-to-max-cycles-times" logic as this loop is extremely large and likely won't perform too well if repeated too much. Would expect rather nasty bumps in individual frame timings.

Would it be possible to instead change it so that if after one iteration if there's updates still pending they'd be deferred until the next game tick - with a blessing to bypass any internal update throttling?

Reply:
I have redesigned the `Update()` partially, some scripts now run outside of the loop.
The pseudo-semaphore is still in to control recursion depth, but the 99% use-case should be one cycle.
Also, order of execution is more predictable now.